src/compiler/scala/tools/nsc/typechecker/Contexts.scala

```diff
@@ -484,7 +484,10 @@ trait Contexts { self: Analyzer =>
       c(ConstructorSuffix) = !isTemplateOrPackage && c(ConstructorSuffix)
 
       // scala/bug#8245 `isLazy` need to skip lazy getters to ensure `return` binds to the right place
-      c.enclMethod         = if (isDefDef && !owner.isLazy) c else enclMethod
+      // scala/bug#10291 ditto for the applyOrElse synthetic method in a PartialFunction
+      // (a PF could also contain an `apply` instead under the @cps plugin, but this is no longer supported)
+      def skipAsEnclosingMethod = owner.isLazy || (owner.name == nme.applyOrElse && owner.owner.isAnonymousFunction)
+      c.enclMethod         = if (isDefDef && !skipAsEnclosingMethod) c else enclMethod
```

Comment:
Could this inadvertently catch user-defined `applyOrElse`?

Is it possible to drive this instead with:

```diff
diff --git a/src/compiler/scala/tools/nsc/typechecker/Typers.scala b/src/compiler/scala/tools/nsc/typechecker/Typers.scala
index 57e46d95d7..32cd15347c 100644
--- a/src/compiler/scala/tools/nsc/typechecker/Typers.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Typers.scala
@@ -2667,7 +2667,9 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
         val paramSyms = List(x, default)
         methodSym setInfo genPolyType(List(A1, B1), MethodType(paramSyms, B1.tpe))

-        val methodBodyTyper = newTyper(context.makeNewScope(context.tree, methodSym))
+        val nestedContext = context.makeNewScope(context.tree, methodSym)
+        nestedContext.enclMethod = context.enclClass
+        val methodBodyTyper = newTyper(nestedContext)
         if (!paramSynthetic) methodBodyTyper.context.scope enter x

         // First, type without the default case; only the cases provided
```

Reply:
good spot! nice opportunity for cleanup -- now, if someone returns from the guard in isDefinedAt, it'll also be the right kind of return :-)